_posts/2024-04-12-vector-search.md

```diff
@@ -0,0 +1,104 @@
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│array_distance(vec, [1.0, 2│
+│         .0, 3.0])         │
+└─────────────┬─────────────┘                             
+┌─────────────┴─────────────┐
+│      HNSW_INDEX_SCAN      │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│   t1 (HNSW INDEX SCAN :   │
+│            idx)           │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│           EC: 3           │
+└───────────────────────────┘   
+```
+
+You can pass the `HNSW` index creation statement a `metric` parameter to decide what kind of distance metric to use. The supported metrics are `l2sq`, `cosine` and `inner_product`, matching the three built in distance functions: `array_distance`, `array_cosine_distance` and `array_inner_product`.
+ The default is `l2sq` (`array_distance`)..
+
+```sql
+-- Create an HNSW index on the column using euclidean distance
+CREATE INDEX l2sq_idx ON embeddings USING HNSW (vec) WITH (metric = 'l2sq');
+
+-- Create an HNSW index on the column using cosine distance
+CREATE INDEX cos_idx ON embeddings USING HNSW (vec) WITH (metric = 'cosine');
+
+-- Create an HNSW index on the column using inner product
+CREATE INDEX ip_idx ON embeddings USING HNSW (vec) WITH (metric = 'ip');
+```
+
+It is important to know that the `HNSW` index is a _approximate_ index, meaning that the results returned by the index may not be the _exact_ top-n results, but rather a set of results that are _close_ to the top-n results. The `HNSW` in `HNSW` index stands for Hierarchical Navigable Small World, which is a graph-based index structure that is particularly well suited for high-dimensional vector similarity search. 
+
+
+While the `HNSW` index supports deletes and insertions to the base table after the index has been created, deletes are not immediately reflected internally in the index structure, but are instead marked as deleted, and will only be removed after a certain number of updates have been made. This is a trade-off to maintain the performance of updates, but will lead to the _quality_ of the index deteriorating over time, impacting both speed and accuracy of retrieval. To remedy this you can use the pragma `pragma_hnsw_compact_index('<index name>')` to trigger a recompaction, which will be faster than dropping and recreating the index. 
+
+Additionally, it is generally recommended to create the index __after__ loading a table with data as the initial index creation step can utilize some degree of parallelism to speed up index creation on larger tables.
+
+
+Like the `ART` the `HNSW` index must be able to fit into RAM in its entirety, but will also be persisted to disk as part of the DuckDB database file (if running DuckDB in disk-backed mode) during checkpointing. After restarting or loading a database file with a persisted `HNSW` index, the index will only be lazily loaded back into memory whenever the table the index is created on is first accessed.
+
```

Comment:
You included some info about the library you used for the index in the docs - do you want to include it in the post as well? I would only include it if you want to sing its praises a bit (Ex: 10x faster than FAISS while embracing DuckDB's zero dependency ethos, C++11 simplicity, etc.). That would fit nicely in a blog, but would read a bit odd in the docs... :-)